# 03 – Feature Engineering

### Purpose
This notebook creates engineered features that enhance the predictive power and interpretability of the TED CAN dataset.
All features are based on insights from Notebook 01 (EDA) and domain knowledge of public procurement.

### Steps
- Load cleaned‑saved datasets
- Apply feature engineering pypline
- Save full datasets

--------------------
#### Imports & Setup & Dataset
-------------------

In [35]:
# ---------------------------------------------------------
# Import moduls
# ---------------------------------------------------------


# import standard modules
import pandas as pd
import numpy as np
from pathlib import Path

import sys
from pathlib import Path

In [36]:
# shut off some annoying warnings
import warnings

warnings.filterwarnings("ignore", message="A value is trying to be set on a copy")
warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=FutureWarning)

In [37]:
# ---------------------------------------------------------
# Setup style
# ---------------------------------------------------------

# show all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# visualisation settings
pd.set_option('display.float_format', '{:,.2f}'.format)

In [38]:
# ---------------------------------------------------------
# Load scripts
# ----------------------------------------------------------

%load_ext autoreload
%autoreload 2

from pathlib import Path
import sys

# found min directory
PROJECT_ROOT = Path("..").resolve()

# maindirectory sys.path
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

# import function from script
from my_scripts.feature_engineering import feature_engineering
from my_scripts.eda import overview

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [39]:
# ---------------------------------------------------------
# Import dataset
# ---------------------------------------------------------

df = pd.read_pickle("../data/dataset_clean.pkl")

print("EU dataset:", df.shape)

EU dataset: (2993666, 14)


In [40]:
df.head()

,YEAR,ISO_COUNTRY_CODE,CAE_TYPE,TYPE_OF_CONTRACT,TOP_TYPE,CPV,MAIN_CPV_CODE_GPA,AWARD_VALUE_EURO,CRIT_PRICE_WEIGHT,NUMBER_OFFERS,DT_DISPATCH,DT_AWARD,TITLE,CRIT_WEIGHTS
0,2008,DE,8,W,OPE,45215000.0,Unknown,"302,964.15",NaN,2.00,2007-12-11,2007-09-29,Unknown,60---40
1,2008,DE,3,W,OPE,45215100.0,Unknown,"478,780.08",100.00,3.00,2007-12-27,2007-12-21,Unknown,Unknown
2,2008,ES,1,W,OPE,45234116.0,Unknown,"14,237,368.10",NaN,6.00,2007-12-27,2007-10-25,Unknown,60---40
3,2008,DE,3,W,OPE,45233110.0,Unknown,"5,204,457.56",NaN,9.00,2007-12-27,2007-11-30,Unknown,90---10
4,2008,ES,3,W,OPE,45233222.0,Unknown,"4,924,202.16",NaN,45.00,2007-12-28,2007-12-14,Unknown,70---30


---------------------------------------------------------
### Apply feature engineering

-----------------------------

In [41]:
# apply funktion
df_fe = feature_engineering(df)


In [42]:
# shape of the dataset
print(df_fe.shape)

(2986847, 15)


In [43]:
# inspekt of the EU dataset
overview(df_fe)

,dtype,total,missing_n,missing_%,uniques_n,uniques
YEAR,int16,2986847,0,0.00,9,"[2008, 2009, 2010, 2011, 2012, 2013, 2014, 201..."
ISO_COUNTRY_CODE,str,2986847,0,0.00,33,"[DE, ES, FR, PL, HU, IT, CY, UK, RO, PT, SE, N..."
TYPE_OF_CONTRACT,str,2986847,0,0.00,3,"[W, U, S]"
TOP_TYPE,str,2986847,0,0.00,10,"[OPE, NIC, RES, NOC, Unknown, COD, AWP, NOP, N..."
AWARD_VALUE_EURO,float64,2986847,0,0.00,1600978,"[302964.15, 478780.08, 14237368.1, 5204457.56,..."
CRIT_PRICE_WEIGHT,float64,2986847,0,0.00,120,"[60.0, 100.0, 90.0, 70.0, 40.0, 30.0, 33.0, 55..."
NUMBER_OFFERS,float64,2986847,0,0.00,361,"[2.0, 3.0, 6.0, 9.0, 45.0, 1.0, 5.0, 4.0, 10.0..."
TITLE,str,2986847,0,0.00,967027,"[Unknown, Pakiet Nr 4., Pakiet Nr 5., Pakiet N..."
IS_FAILED_TENDER,int8,2986847,0,0.00,2,"[0, 1]"
AWARD_QUARTER,float64,2986847,0,0.00,4,"[3.0, 4.0, 1.0, 2.0]"


In [44]:
df_fe.describe().T

,count,mean,std,min,25%,50%,75%,max
YEAR,"2,986,847.00","2,012.35",2.53,"2,008.00","2,010.00","2,012.00","2,015.00","2,016.00"
AWARD_VALUE_EURO,"2,986,847.00","6,235,406.57","91,555,681.20",0.01,"6,389.49","63,909.83","411,549.72","9,999,999,999.00"
CRIT_PRICE_WEIGHT,"2,986,847.00",83.43,25.78,0.00,60.00,100.00,100.00,100.00
NUMBER_OFFERS,"2,986,847.00",7.02,18.61,0.00,1.00,3.00,6.00,500.00
IS_FAILED_TENDER,"2,986,847.00",0.26,0.44,0.00,0.00,0.00,1.00,1.00
AWARD_QUARTER,"2,986,847.00",2.60,1.12,1.00,2.00,3.00,4.00,4.00
DAYS_TO_AWARD,"2,986,847.00",50.05,78.46,0.00,13.00,30.00,55.00,"1,809.00"
DAYS_TO_AWARD_MISSING,"2,986,847.00",0.00,0.04,0.00,0.00,0.00,0.00,1.00
CRIT_PRICE_WEIGHT_MISSING,"2,986,847.00",0.51,0.50,0.00,0.00,1.00,1.00,1.00


#### Notes: Feature Engineering Summary
Feature engineering enriches the cleaned dataset with analytical variables that improve modelling of tender competition, procedural behaviour, thematic structure (NLP), and risk indicators.

1. Dataset Size After Feature Engineering
- Before: 2,993,666 rows × 14 columns  
- After: 2,986,847 rows × 15 columns  
(≈ 0.23% rows removed due to unrealistic or missing reconstructed values)

2. Created Features
- AWARD_QUARTER — quarter of the award date  
- DAYS_TO_AWARD — duration between dispatch and award  
  - Negative durations removed  
  - Durations > 5 years (1825 days) removed  
- DAYS_TO_AWARD_MISSING — indicator for unrealistic or missing durations  

These features capture temporal patterns, seasonality, and procedural duration while ensuring logical consistency.

- CPV 12‑group mapping (CPV_CATEGORY_12)
Based on CPV divisions:
  - Construction & Building Works: 45–46  
  - Machinery & Industrial Equipment: 30–32
  - Electrical, Optical & Precision Equipment: 33–34 
  - IT, Software & Telecommunications: 48–50  
  - Transport Services & Logistics: 60–63  
  - Business, Consulting & Administrative Services: 79–80  
  - Cleaning, Facility & Maintenance Services: 90–91  
  - Energy, Utilities & Environmental Services: 92–99  
  - Other
This mapping provides interpretable thematic categories for modelling and dashboards.

- CAE Type Mapping
Categorisation of contracting authorities:
  - 1 - Ministry  
  - 3 - National Agency  
  - 4 - Regional/Local Authority  
  - 5 - Regional/Local Agency  
  - 5A - Local Agency  
  - 6 - Public Body  
  - 8 - Public Undertaking  
  - N - Not Specified  
  - R - Other Public Authority  
  - Z - Unknown  

- Missing‑information flags
  - CRIT_PRICE_WEIGHT_MISSING — missing or reconstructed values  
  - DAYS_TO_AWARD_MISSING — unrealistic or missing date differences  
These indicators allow the model to capture informative missingness.

3. Reconstructed Features
- CRIT_PRICE_WEIGHT
Reconstructed using:
  - Original numeric values  
  - If missing → extracted from "CRIT_WEIGHTS" (first 1–3 digits of patterns like "60---40")  
  - Unrealistic values > 100 removed  
  - Remaining NaN imputed with median  

This recovers ~5% additional valid observations.

4. Numeric Missing Values
All numeric columns with NaN are imputed using **median**, ensuring stability and robustness.

5. Removed Raw Columns
To avoid redundancy after feature engineering:
- "DT_AWARD", "DT_DISPATCH"  
- "MAIN_CPV_CODE_GPA"  
- "CAE_TYPE"  
- "CRIT_WEIGHTS"
These fields were replaced by engineered features.



------------------
### Save engineered datasets

-------------------

In [45]:
df_fe.to_pickle("../data/dataset_fe.pkl")